hathor/transaction/storage/block_height_index.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2021 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Optional, Tuple
+
+from hathor.transaction.genesis import BLOCK_GENESIS
+from hathor.util import not_none
+
+BLOCK_GENESIS_HASH: bytes = not_none(BLOCK_GENESIS.hash)
+
+
+class BlockHeightIndex:
+    """Store the block hash for each given height
+    """
+    def __init__(self) -> None:
+        self._index: List[bytes] = [BLOCK_GENESIS_HASH]
+
+    def add(self, height: int, block_hash: bytes) -> bool:
+        """Add or update an element in the cache. Deals with the corner cases so the caller doesn't have to.
+
+        Returns False if the new element means there was a reorg and True otherwise.
+        """
+        if len(self._index) < height:
+            raise ValueError(f'parent hash required (current height: {len(self._index)}, new height: {height})')
+        elif len(self._index) == height:
+            self._index.append(block_hash)
+        elif self._index[height] != block_hash:
```

Comment:
Should we do it in a separate method? Something like `add_and_reorg(...)`? So this method would fail if it tries to update with a reorg.

Reply:
Makes sense.

Reply:
I'm splitting it into `add_new` and `add_reorg`, what do you think?

Reply:
Done.